src/com/google/javascript/jscomp/RewriteDynamicImports.java

```diff
@@ -357,4 +375,82 @@ private void injectWrappingFunctionExtern() {
     externsRoot.addChildToBack(wrappingFunctionDefinition);
     compiler.reportChangeToEnclosingScope(wrappingFunctionDefinition);
   }
+
+  /**
+   * A shallow traversal class for the externs to inject the alias.
+   *
+   * Later passes require the names to be defined. For simple name aliases first check
+   * for an existing definition. Qualified name aliases check for an existing definition of the
+   * root name only.
+   */
+  private class AliasInjectingTraversal extends NodeTraversal.AbstractPreOrderCallback {
+    @Override
+    public boolean shouldTraverse(NodeTraversal t, Node n, Node parent) {
+      checkState(aliasIsValid());
+
+      // If the extern is a simple name and already declared, there is nothing to do
+      if (NodeUtil.isValidSimpleName(alias) && t.getScope().getVar(alias) != null) {
```

Comment:
I'm suspicious that all of this work to get the types is for naught.
I'm pretty sure this pass is running before the type checking passes.
Check DefaultPassConfig.

I think you'd probably be better off to just inject some parsed code with appropriate JS doc.

Something like this

https://github.com/google/closure-compiler/blob/master/src/com/google/javascript/jscomp/CrossChunkMethodMotion.java#L139

Or else do the injection after type checking when the types will be present.

Reply:
It's definitely running after type checking:

 - [check types](https://github.com/google/closure-compiler/blob/7ce5df09a638fa1e31f3407b0d2db71fc20ae8f6/src/com/google/javascript/jscomp/DefaultPassConfig.java#L353)
 - [rewrite dynamic imports](https://github.com/google/closure-compiler/blob/7ce5df09a638fa1e31f3407b0d2db71fc20ae8f6/src/com/google/javascript/jscomp/DefaultPassConfig.java#L381)
 
It is more readable parsing synthetic code, but later passes then crash because they assume type information is present and it isn't.